text/0000-std-rpc-api.md

```diff
@@ -0,0 +1,393 @@
+- Feature Name: `std-rpc-api`
+- Start Date: 2020-11-02
+- RFC PR: [Unity-Technologies/com.unity.multiplayer.rfcs#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/0000)
+- Issue: [Unity-Technologies/com.unity.multiplayer#0000](https://github.com/Unity-Technologies/com.unity.multiplayer/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Existing MLAPI [**Convenience** and **Performance** RPCs](https://mlapi.network/wiki/messaging-system/) are sub-optimal in terms of required boilerplate code, API design and runtime performance.
+
+This RFC proposes a new **Standard** RPC API that excels at minimal boilerplate, API design/clarity, extensibility (future-proofing) and runtime performance.
+
+The initial implementation of this standard would be kept minimal and further-extended with additional features down the line as it will also support future extensions and enhancements without changing the standard API necessarily.
+
+# Motivation
+[motivation]: #motivation
+
+Achieve better performance, unify & standardize RPC API, make RPC API extensible & future-proof and reduce boilerplate code.
+
+Also a simple internal performance benchmark came with these results:
+
+    Local RPC Send Performance Test (524280 calls)
+    Convenience -> 4361ms
+    Performance -> 2607ms
+    Standard -> 2205ms
+    Local RPC Recv Performance Test (524280 calls)
+    Convenience -> 9626ms
+    Performance -> 5923ms
+    Standard -> 1935ms
+
+We're looking at **up to 5.24x better performance** (compared to convenience API vs standard API receive performance).
+
+Beyond that, on an internal project, we saw roughly **%80 less boilerplate code** compared to performance RPCs and standard RPCs which is also a huge win.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The multiplayer framework provides 2 main network constructs ([ServerRpc](#serverrpc) and [ClientRpc](#clientrpc)) to execute logic on either server-side or client-side. This concept is often called [Remote Procedure Call (RPC)](https://en.wikipedia.org/wiki/Remote_procedure_call) and has wide adoption across the industry.
+
+## RPC Methods
+
+A framework user (Unity developer) can declare multiple RPCs under a `NetworkBehaviour` and inbound/outbound RPC calls will be replicated as a part of its replication in a network frame.
+
+A method turned into an RPC is no longer a regular method, it will have its  own implications on direct calls and in the network pipeline (see [Execution Table](#execution-table)).
+
+### ServerRpc
+
+A `ServerRpc` can be invoked by a client to be executed on the server.
+
+Developer can declare a `ServerRpc` by marking a method with `[ServerRpc]` attribute and making sure to have `ServerRpc` suffix in the method name.
+
+```cs
+[ServerRpc]
+void PingServerRpc(int framekey) { /* ... */ }
```

Comment:
How can this setup work with a, as MLAPI describes it, [multiple project setup](https://mlapi.network/wiki/messaging-system/)? E.g. my server is a completely different Unity project, the `server` methods dont exist in the client project and vice-versa. In MLAPI i can invoke methods by string parameter.

Maybe a similar scenario can be made in the same project, where in order to separate concerns i split Server and Client up in different scripts. The Server adds the `ServerManageGameState` NetworkedBehaviour, the client adds the `ClientManageGameState` NetworkedBehaviour. 

How can these two scripts send RPCs to eachother?

Reply:
There is no cross-binary (and also no cross-version) compatibility in MLAPI today. Also an RPC can't map to another object. What you can do is though, have a `NetworkBehaviour` full of `ClientRpc`s and `ServerRpc`s, then call `ClientRpc`s inside `ServerRpc`s and that'd just work fine.